src/NuGet.Core/NuGet.Build.Tasks.Console/Program.cs

```diff
@@ -144,23 +147,62 @@ private static bool IsDebug()
         /// <returns><code>true</code> if the arguments were successfully parsed, otherwise <code>false</code>.</returns>
-        private static bool TryParseArguments(string[] args, out (Dictionary<string, string> Options, FileInfo MSBuildExeFilePath, string EntryProjectFilePath, Dictionary<string, string> MSBuildGlobalProperties) arguments)
+        internal static bool TryParseArguments(string[] args, Func<Stream> getStream, TextWriter errorWriter, out (Dictionary<string, string> Options, FileInfo MSBuildExeFilePath, string EntryProjectFilePath, Dictionary<string, string> MSBuildGlobalProperties) arguments)
         {
-            if (args.Length != 4)
+            arguments = (null, null, null, null);
+
+            // This application receives 3 or 4 arguments:
+            // 1. A semicolon delimited list of key value pairs that are the options to the program
+            // 2. The full path to MSBuild.exe
+            // 3. The full path to the entry project file
+            // 4. (optional) A semicolon delimited list of key value pairs that are the global properties to pass to MSBuild
+            if (args.Length < 3 || args.Length > 4)
             {
-                arguments = (null, null, null, null);
+                // An error occurred parsing command-line arguments in static graph-based restore as there was an unexpected number of arguments, {0}. Please file an issue at https://github.com/NuGet/Home. {0}
+                LogError(errorWriter, Strings.Error_StaticGraphRestoreArgumentParsingFailedInvalidNumberOfArguments, args.Length);
 
                 return false;
             }
 
             try
             {
-                var options = ParseSemicolonDelimitedListOfKeyValuePairs(args[0]);
+                Dictionary<string, string> options = ParseSemicolonDelimitedListOfKeyValuePairs(args[0]);
                 var msbuildExeFilePath = new FileInfo(args[1]);
                 var entryProjectFilePath = args[2];
-                var globalProperties = ParseSemicolonDelimitedListOfKeyValuePairs(args[3]);
+
+                Dictionary<string, string> globalProperties = null;
+
+                // If there are 3 arguments then the global properties will be read from STDIN
+                if (args.Length == 3)
+                {
+#if NETFRAMEWORK
+                    if (AppDomain.CurrentDomain.IsDefaultAppDomain())
+                    {
+                        // The application is called twice and the first invocation does not need to read the global properties
+                        globalProperties = null;
+                    }
+                    else
+                    {
+#endif
+                        using var reader = new BinaryReader(getStream(), new UTF8Encoding(encoderShouldEmitUTF8Identifier: false, throwOnInvalidBytes: true), leaveOpen: true);
```

Comment:
The caller is expected to close the stream rather than have this reader close it.  The caller could be production code or a unit test and I wanted the contract to be that this method left the stream open.